spec/formatting.md

```diff
@@ -19,6 +19,112 @@ the local variable takes precedence.
 It is an error for a local variable definition to
 refer to a local variable that's defined after it in the message.
 
+## Pattern Selection
+
+When formatting a message with one or more selectors,
+the Pattern of one of Variants must be selected for formatting.
+
+An implementation may use any pattern selection method,
+as long as its observable behaviour matches the results of the method defined here.
+
+### Setup
+
+To select a Pattern,
+first the resolved values _res_ of the Selector must be determined:
+
+1. Let _res_ be a new empty list of resolved values that support selection.
+2. For each Expression _exp_ of the message's Selector Expressions,
+   1. Let _rv_ be the resolved value of _exp_.
+   2. If selection is supported for _rv_:
+      1. Append _rv_ as the last element of the list _res_.
+   3. Else:
+      1. Emit a Selection Error.
+      2. Let _nomatch_ be a resolved value for which selection always fails.
+      3. Append _nomatch_ as the last element of the list _res_.
+
+The shape of the resolved values must be determined by each implementation,
+along with the manner of determining their support for selection.
```

Comment:
I'd like to push back a little on formatting being an explicit requirement for selection. In many cases I agree that formatting is an implicit practical requirement, but that's an implementation detail. The CLDR rules you link to depend on [Plural Rule Operands](http://unicode.org/reports/tr35/tr35-numbers.html#Operands), which at least in JS [are calculated](https://tc39.es/ecma402/#sec-resolveplural) from a locale-independent formatting of the input value that's relatively easy to parse. For example, digit grouping is not done and the decimal digit is always a period `.`.

Theoretically, it would be possible to determine these operand values without the intermediate formatting step, and in certain cases it might be possible to reuse the formatting if it happens to match the expected output for the current locale and options. But in practice the reasonable thing to do is to format & re-parse the input number within the plural selector, and separately format the input number for string output.

With the French millions case, for instance, the plural selection ends up depending on the formatting to `1.2M`, while the formatting of a matching placeholder would end up as `1,2 M`.

Reply:
> I'd like to push back a little on formatting being an explicit requirement for selection. In many cases I agree that formatting is an implicit practical requirement, but that's an implementation detail. ... [PluralRuleOperands] at least in JS [are calculated](https://tc39.es/ecma402/#sec-resolveplural) from a locale-independent formatting of the input value that's relatively easy to parse. For example, digit grouping is not done and the decimal digit is always a period `.`. ...  With the French millions case, for instance, the plural selection ends up depending on the formatting to `1.2M`, while the formatting of a matching placeholder would end up as `1,2 M`.

We might be partially talking past each other due to overloaded terminology, and there's still an unaddressed point.

Formatting usually means returning a string, but sometimes in ICU when formatting type `X`, an intermediate representation after applying some locale-specific processing is sometimes called `FormattedX`. This intermediate pre-processed state still occurs before the formatting symbols are applied when getting the final formatted string. (ex: [`FormattedNumber`](https://unicode-org.github.io/icu-docs/apidoc/dev/icu4j/com/ibm/icu/number/FormattedNumber.html) is intermediate, it still has a `toString()` method, but can also be an input to selection in `PluralRules.select()`). Instead of using a string adhering to a grammar like in JS, ICU4X uses a more structured type ([`FixedDecimal`](https://icu4x.unicode.org/doc/fixed_decimal/index.html)) for the intermediate value that can still be used for [selection](https://icu4x.unicode.org/doc/icu_plurals/struct.PluralRules.html#method.category_for) and [formatting to string](https://icu4x.unicode.org/doc/icu_decimal/struct.FixedDecimalFormatter.html). That avoids the redundant need to parse strings to reconstruct that information if using the JS string approach.

This preprocessing step is still usually handled by a formatter, so we still have that as a dependency in this example. In practice, the logic for selection is going to be closely related to the logic for formatting. These things are often intertwined. The other higher level point that still seems unaccounted for is that _how the selection occurs_ is non trivial (it's not string equality). 

The spec still should be specific on who or how that selection should be done. As it stands currently, it's underspecified.  And the text instead asserts to the effect that "the implementation will determine ...", which implies the MF2 implementation. But that's not the appropriate level of responsibility for matching / selection, which should be closer to the formatting.

Reply:
> We might be partially talking past each other due to overloaded terminology

Yeah, it's sounding like what I'm referring to as a "resolved value" could be represented in ICU4J by the `FormattedX` entities. At least in my headcanon, the whole process of what could be called "formatting" makes more sense to split into two: "resolution" and "formatting". In the first, you gather up all the information you need in order to do e.g. formatting or selection, and in the second you take that information and you emit a value in the final representation that you need.

> [...] In practice, the logic for selection is going to be closely related to the logic for formatting. These things are often intertwined. [...]

Definitely agree, for plural selection.

> The spec still should be specific on who or how that selection should be done. As it stands currently, it's underspecified. And the text instead asserts to the effect that "the implementation will determine ...", which implies the MF2 implementation. But that's not the appropriate level of responsibility for matching / selection, which should be closer to the formatting.

While I agree that the behaviour of a plural selector (be it `:number` or `:plural`) must be well defined for use in an implementation, I do not think that definition belongs in _this specification_.

Do you think we ought to include something like what I mention in [this comment](https://github.com/unicode-org/message-format-wg/pull/333/files#r1094217500) in the MF2 spec, though? As I understand it, a FormattedNumber should work as an implementation of a "MessageNumber".

Reply:
> in my headcanon ...  In ["resolution"], you gather up all the information you need in order to do e.g. formatting or selection...

As far as confusing terminology goes, "resolution" still sounds a little too vague. This initial preprocessing step during the overall work of "formatting" (input value -> string) also depends on locale information too. Ex: for compact notation numbers, the exponent that you use is informed by the grouping strategy for numbers in the locale.  (@eggrobin, did I get that example right?)

> > The spec still should be specific on who or how that selection should be done. As it stands currently, it's underspecified. And the text instead asserts to the effect that "the implementation will determine ...", which implies the MF2 implementation. But that's not the appropriate level of responsibility for matching / selection, which should be closer to the formatting.
> 
> While I agree that the behaviour of a plural selector (be it `:number` or `:plural`) must be well defined for use in an implementation, I do not think that definition belongs in _this specification_.

To clarify, I'm not saying that we should define the behavior of plural selector in the spec. What I'm saying is that this PR codifies an algorithm for a first match strategy, but the notion of "match" between a selector value and variant key is underspecified. We are in agreement that the notion of match has to be implemented separately for each type of selector, and those implementation details are not a concern for the spec text. But what I am saying is that we also clearly can't leave the story at that, since the proposed algorithm in spec text is built on top of an assumption of a specific notion of match (equality, ex: string equality), and we know from the plural selector example that that specific notion is insufficient to cover all cases, so it needs to be generalized.

And so, I think we already agree that there is a clear connection there between the proposed high-level algorithm for variant key selection in MF2 and the selector-specific notion of match. In order to resolve the problem that I described above, just as we have done before when designing for things like formatting functions, the proper way to achieve both:

1. generalizing the notion of matching from simple string equality comparison predicate function to an impl-specific predicate function
2. decoupling those impl-specific notions of matching from the high-level algorithm
 
...is to have an interface representing the selection predicate function.

That is the specificity that the current proposed algorithm needs. Without doing so, the algorithm text in the PR here will ignore a design problem that we know that we have. Having an interface to represent a specific selector's impl-specific behavior achieves proper simplicity through decoupling that makes for a good design around this problem.


> Do you think we ought to include something like what I mention in [this comment](https://github.com/unicode-org/message-format-wg/pull/333/files#r1094217500) in the MF2 spec, though? As I understand it, a FormattedNumber should work as an implementation of a "MessageNumber".

I think what you were mentioning in that comment is the same as what I have been saying so far with "selection depends on formatting". But maybe that wasn't clear because we used different words to describe the same structured pre-processed form ("formatting" as in `.formatToParts()` in EMCA-402 vs. "resolved value").

The way you describes it makes me unsure yet if we agree on how best to design for this. Rather than trying to specify the _type_ of the value in the pre-processed structured form (can we call this form the "Preformatting Structured Parts" maybe?), I think we could once again define an _interface for the function_ that returns the value. That would allows us to decouple separate concerns cleanly and simply. But I think we agree in principle, if I understand correctly.

And it would help us all to eventually find some consistent precise naming, too.